staging/src/k8s.io/client-go/features/testing/features.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testing
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"testing"
+
+	clientfeatures "k8s.io/client-go/features"
+)
+
+var (
+	lock sync.Mutex
+)
+
+type dereferencingDelegate interface {
+	DereferenceDelegate() clientfeatures.Gates
+}
+
+// SetFeatureDuringTest sets the specified feature to the specified value for the duration of the test.
+//
+// Example use:
+//
+//	clientfeaturestesting.SetFeatureDuringTest(t, clientfeatures.WatchListClient, false)
+func SetFeatureDuringTest(tb testing.TB, feature clientfeatures.Feature, featureValue bool) {
+	lock.Lock()
+	defer lock.Unlock()
+
+	if err := setFeatureDuringTestInternal(tb, feature, featureValue); err != nil {
+		tb.Fatal(err)
+	}
+}
+
+func setFeatureDuringTestInternal(tb testing.TB, feature clientfeatures.Feature, featureValue bool) error {
+	currentFeatureGates := clientfeatures.FeatureGates()
+	dereferencedFeatureGates, ok := currentFeatureGates.(dereferencingDelegate)
+	if !ok {
+		panic("clientfeatures.FeatureGates() does not implement DereferencingDelegate method")
+	}
+
+	currentFeatureGatesOverrides, ok := dereferencedFeatureGates.DereferenceDelegate().(*featureGatesOverrides)
+	if !ok {
+		// consider adding a mechanism that would detect
```

Comment:
Are we doing all of this to avoid adding a `Set` method?

Reply:
Yeah, why replace instead of adding a Set method? https://github.com/kubernetes/kubernetes/pull/124780/files#diff-92230226e2552b2a77baffee293c03ca22d6a0536aa93116227a16b0fd55d23cR102

Reply:
> Are we doing all of this to avoid adding a Set method?

Not exactly. The Set method does not solve the problem that another test might use Replace. If we had a Set method, the detection mechanism would probably look different.

> why replace instead of adding a Set method?

please see https://github.com/kubernetes/kubernetes/pull/123974#discussion_r1571008742

Reply:
OK, added the `Set` method to the default impl, PTAL - https://github.com/kubernetes/kubernetes/pull/125052
Once that PR merges, the e2e changed it this PR will pass and then we could merge this PR.